### ISSUE TITLE
how does memory-efficient-fp16 work?

### ISSUE THREAD

**Issue by LSC527** [pytorch/fairseq#3751]
#### What is your question?
I don't understand how can memory-efficient-fp16 complete fp16 training only by maintaining the fp32 gradients in optimizer. Is there a detailed description about it？
https://github.com/pytorch/fairseq/blob/237184e5222b347475456f4a44f31a510c64ca35/fairseq/optim/fp16_optimizer.py#L456
